### ISSUE TITLE
[BUG] DTensor RNG tracker is applied to CPU tensors under FSDP2 CPUOffloadPolicy, resetting the CPU RNG state

### ISSUE THREAD

**Issue by francesco-bertolotti** [pytorch/pytorch#196562]
## 🐛 Describe the bug

When a module is wrapped with `fully_shard(..., offload_policy=CPUOffloadPolicy())`, its parameters keep a CUDA-typed `DeviceMesh` even though the local shards live on CPU. Random ops on those parameters are still dispatched through the process-wide DTensor RNG tracker, which is configured for CUDA.

The tracker's `_distribute_region()` enters

```python
torch.random.fork_rng(devices=[cuda_device], ...)
```

and `fork_rng` restores the **CPU** default generator state on exit regardless of which accelerator devices are listed. Every CPU-side random op routed through the mismatched tracker therefore starts from the same CPU RNG state and produces the same values.

Two consequences:

1. **Silently identical values.** Parameters initialized after being offloaded to CPU can receive the same values on every call — same shard content across calls, and identical draws where independent ones are expected. This is silent: no error, no warning, just correlated initialization.
2. **Hangs in rejection samplers.** `nn.init.trunc_normal_` draws from `normal_` and resamples anything outside the requested range. If `normal_` keeps returning the same out-of-range values, the loop never makes progress and initialization hangs indefinitely. This is how the bug was originally found, in torchtitan.

A secondary problem in the same path: the tracker-managed branch pops the `generator` kwarg before it decides whether to handle the op, so a caller-provided `generator` can be dropped on the fallback path.

### Minimal reproduction

Reproduced on a GH200, single rank:

```python
import torch
import torch.distributed as dist
from torch import nn
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.fsdp import CPUOffloadPolicy, fully_shard


@torch.no_grad()
def main():
    assert torch.cuda.is_available(), "this repro requires a CUDA device"
    dist.init_process_group(backend="gloo")

    dp_mesh = init_device_mesh("cuda", (1,))

    lin = nn.Linear(4, 4, bias=False)
    fully_shard(lin, mesh=dp_mesh, offload_policy=CPUOffloadPolicy())

    lin.to_empty(device="cpu")

    for i in range(3):
        lin.weight.normal_()
        print(f"iter {i}: {lin.weight.to_local().flatten()}")

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

Run with `torchrun --nproc-per-node=1 repro.py`.

**Expected:** three different tensors.

**Actual:** all three iterations print the same values.

Swapping `normal_()` for `nn.init.trunc_normal_(lin.weight, a=-2, b=2)` hangs instead of returning.

### Root cause

`torch.distributed.tensor._random._rng_tracker` is a process-wide singleton and carries a single `_device`. The dispatch path in `torch/distributed/tensor/_dispatch.py` decides to use the tracker based on the op being random and a tracker existing — it does not check that the local tensor actually lives on the tracker's device. Under CPU offload those two diverge, and the CUDA-configured tracker ends up wrapping CPU ops.

### Proposed fix

Both changes are in the tracker-dispatch path in `torch/distributed/tensor/_dispatch.py`:

1. Only take the tracker-managed path when the local tensor is on the tracker's device, i.e. gate on `first_local_arg.device.type == random._rng_tracker._device.type`. Otherwise fall through to the normal local-op path.
2. Restore a caller-provided `generator` kwarg before calling `op_call` on the fallback path, so it is not silently dropped after being popped.

A regression test (`DTensorCPUOffloadRandomOpsTest` in `test/distributed/tensor/test_random_ops.py`) reproduces the FSDP2 CPU-offload setup and asserts that three consecutive `.normal_()` calls produce different values.

Fix proposed in #196072.

### Open questions

- Is per-device gating the right layer, or should the tracker itself be device-aware (e.g. a tracker per device type, or the offloaded local tensor's device reflected in the mesh)?
- How broadly does this affect torchtitan? Only initialization is confirmed so far; any random op on an offloaded parameter (dropout on offloaded activations, resampling-based init) is potentially affected.

## Versions

```
torch: 2.15.0.dev20260902+cu130
```

## Related

- Fix: https://github.com/pytorch/pytorch/pull/196072
- Downstream report: https://github.com/pytorch/torchtitan/issues/4511

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @weifengpy @kapilsh @zhaojuanmao @mrshenli @rohan-varma @chauhang @mori360 @ppwwyyxx @tianyu-l @XilunWu @SherlockNoMad @albanD @ezyang @pianpwk

### Versions

```
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCm SDK used to build PyTorch: N/A
HIP used to build PyTorch: N/A

OS: Red Hat Enterprise Linux 9.8 (Plow) (aarch64)
GCC version: (GCC) 14.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.34

Python version: 3.9.25 (main, Aug  6 2026, 00:00:00)  [GCC 11.5.0 20240719 (Red Hat 11.5.0-14)] (64-bit runtime)
Python platform: Linux-5.14.0-687.41.1.el9_8.aarch64+64k-aarch64-with-glibc2.34
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GH200 480GB
Nvidia driver version: 595.71.05
cuDNN version: Could not collect
Is XPU available: N/A
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A
Caching allocator config: N/A

CPU:
Architecture:                            aarch64
CPU op-mode(s):                          64-bit
Byte Order:                              Little Endian
CPU(s):                                  72
On-line CPU(s) list:                     0-71
Vendor ID:                               ARM
Model name:                              Neoverse-V2
Model:                                   0
Thread(s) per core:                      1
Core(s) per socket:                      72
Socket(s):                               1
Stepping:                                r0p0
Frequency boost:                         disabled
CPU(s) scaling MHz:                      100%
CPU max MHz:                             3447.0000
CPU min MHz:                             81.0000
BogoMIPS:                                2000.00
Flags:                                   fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm jscvt fcma lrcpc dcpop sha3 sm3 sm4 asimddp sha512 sve asimdfhm dit uscat ilrcpc flagm sb dcpodp sve2 sveaes svepmull svebitperm svesha3 svesm4 flagm2 frint svei8mm svebf16 i8mm bf16 dgh
L1d cache:                               4.5 MiB (72 instances)
L1i cache:                               4.5 MiB (72 instances)
L2 cache:                                72 MiB (72 instances)
L3 cache:                                114 MiB (1 instance)
NUMA node(s):                            9
NUMA node0 CPU(s):                       0-71
NUMA node1 CPU(s):
NUMA node2 CPU(s):
NUMA node3 CPU(s):
NUMA node4 CPU(s):
NUMA node5 CPU(s):
NUMA node6 CPU(s):
NUMA node7 CPU(s):
NUMA node8 CPU(s):
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; CSV2, BHB
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] No relevant packages
[conda] Could not collect
```
